### PR TITLE
feat: make specific fields in HookContext immutable

### PR DESCRIPTION
**Pull request by gruebel**
## This PR

- also added tests to cover section 4.1

### Related Issues

Relates #2

### CODE REVIEW

File: openfeature/hook/__init__.py [hunk 1]
```diff
@@ -29,6 +29,11 @@ class HookContext:
     client_metadata: typing.Optional[ClientMetadata] = None
     provider_metadata: typing.Optional[Metadata] = None
 
+    def __setattr__(self, key: str, value: typing.Any) -> None:
+        if hasattr(self, key) and key in ("flag_key", "flag_type", "default_value"):
```
Comment (federicobond) [inline openfeature/hook/__init__.py:33]: Do we need the `hasattr(self, key)` check? Or can we do with just the `in` expression?

File: openfeature/hook/__init__.py [hunk 2]
```diff
@@ -29,6 +29,11 @@ class HookContext:
...
+            raise AttributeError(f"Attribute {key!r} is immutable")
```
Comment (federicobond) [inline openfeature/hook/__init__.py:34]: What's the difference with just `{key}` here?

Comment (gruebel) [reply openfeature/hook/__init__.py:34]: it just adds quotes around it 😄 ex. `Attribute 'flag_key' is immutable`

File: openfeature/hook/__init__.py (hunk 1 above)
Comment (gruebel) [reply openfeature/hook/__init__.py:33]: yeah, I tried it without it first, but then it fails on the initialization of the dataclass instance.

Comment (federicobond) [reply openfeature/hook/__init__.py:33]: ah makes sense

**Review by federicobond — APPROVED** [review]

**Pull request merged**